src/v/raft/mux_state_machine.h

```diff
@@ -281,17 +281,22 @@ ss::future<> mux_state_machine<T...>::apply(model::record_batch b) {
           *state);
 
         return result_f.then([this, last_offset](std::error_code ec) {
-            auto f = _mutex.with([this, last_offset, ec] {
-                if (auto it = _promises.find(last_offset);
-                    it != _promises.end()) {
-                    it->second.set_value(ec);
-                }
-            });
-            if (!_persist_last_applied) {
-                return f;
-            }
-            return f.then(
-              [this, last_offset] { return write_last_applied(last_offset); });
+            return _mutex
+              .with([this, last_offset, ec] {
+                  if (auto it = _promises.find(last_offset);
+                      it != _promises.end()) {
+                      it->second.set_value(ec);
+                  }
+              })
+              .then([this, last_offset] {
+                  if (
+                    _persist_last_applied
+                    && last_offset > _c->read_last_applied()) {
+                      (void)ss::with_gate(_gate, [this, last_offset] {
+                          return _c->write_last_applied(last_offset);
```

Comment:
it looks like you can have two inflight background kvstore writes where the one with a smaller last offset or the larger one win the race. is that problematic? if the smaller one wins and overwrites the larger one, does that simply mean that extra data will be processed on restart?

Reply:
exactly, this is only an optimization, even if updates will get reordered we will just have to wait for committed offset to be updated.

Reply:
I have the same concern as Noah, when the smaller wins it invalidates [this](https://github.com/vectorizedio/redpanda/commit/b1d1f469b32ebbd210d67282f14def2d0fe59fd5) commit, unfortunately the commit doesn't include info on why `stm.wait` was added so it's unclear to me if it's safe

Reply:
synced via zoom, there is a follow up to add cas to kvstore to prevent out of order updates

Reply:
I went through the code of `storage::kvstore` one again. There is no scheduling points between calling `put` and the time when operation is queued in key value store. In other words writing last applied offset will not be reorderd. I will move dispatching write last applied request in the mutex. This way updates order will be maintained.